test/test_torch.py

```diff
@@ -17485,27 +17485,75 @@ def test_rdiv(self, device, dtype):
         z = torch.tensor([30 / v.item() for v in x], device=device)
         self.assertEqual(y, z, exact_dtype=False)
 
-    @onlyCPU
+    # TO DO
+    # Cross product dtypes after fix type promotion
     @dtypes(*torch.testing.get_all_dtypes(include_bfloat16=False, include_bool=False, include_complex=False))
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
     def test_fmod(self, device, dtype):
-        m1 = torch.Tensor(10, 10).uniform_(-10., 10.).to(dtype=dtype, device=device)
-        res1 = m1.clone()
-        q = 3
-        res1[:, 3].fmod_(q)
-        res2 = m1.clone()
-        for i in range(m1.size(1)):
-            res2[i, 3] = math.fmod(res2[i, 3], q)
-        self.assertEqual(res1, res2)
+        # Use numpy as reference
+        def _reference_implementation(dtype, x, mod):
+            res = torch.fmod(x, mod)
+            # No type promotion
+            # Issue #47779: https://github.com/pytorch/pytorch/issues/47779
+            if torch.is_tensor(mod):
+                if dtype in torch.testing.get_all_int_dtypes():
+                    mod = mod.to(dtype)
+                mod = mod.cpu().numpy()
+            else:
+                if dtype in torch.testing.get_all_int_dtypes():
+                    mod = int(mod)
+            np_x = x.cpu().numpy()
+            exp = np.fmod(np_x, mod)
+            exp = torch.from_numpy(exp)
+            res = res.to(exp.dtype)
+            self.assertEqual(res, exp)
+
+        x = make_tensor((10, 10), device=device, dtype=dtype, low=-9, high=9)
+        # non-contiguours
+        xt = x.t()
+        # Exclude 0
+        mod = make_tensor((10, 10), device=device,
+                          dtype=torch.float if dtype in torch.testing.get_all_int_dtypes() else dtype,
+                          low=1, high=9)
+        zero = torch.zeros_like(x)
+
+        # Integer Scalar
+        _reference_implementation(dtype, x, 3)
+        _reference_implementation(dtype, xt, 3)
+        # Float Scalar
+        _reference_implementation(dtype, x, 2.3)
+        _reference_implementation(dtype, xt, 2.3)
```

Comment:
Could we add some tests for:

- fmod(tensor, mod) where tensor and mod both have the same dtype
- fmod(non-contiguous tensor, non-contiguous mod) where tensor and mod have the same dtype

I'm pretty sure that TensorIterator handles all of these correctly, but it's nice to have tests to prevent future regressions

Reply:
Makes sense.